vm/src/obj/objdict.rs

```diff
@@ -174,231 +173,146 @@ fn dict_new(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
+            .borrow()
+            .values()
+            .map(|(k, v)| vm.ctx.new_tuple(vec![k.clone(), v.clone()]))
+            .collect();
+        let items_list = vm.ctx.new_list(items);
 
-    let iter_obj = PyObject::new(
         PyIteratorValue {
             position: Cell::new(0),
             iterated_obj: items_list,
-        },
-        vm.ctx.iter_type(),
-    );
-
-    Ok(iter_obj)
-}
-
-fn dict_setitem(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    arg_check!(
-        vm,
-        args,
-        required = [
-            (dict, Some(vm.ctx.dict_type())),
-            (needle, Some(vm.ctx.str_type())),
-            (value, None)
-        ]
-    );
-
-    set_item(dict, vm, needle, value);
-
-    Ok(vm.get_none())
-}
-
-fn dict_getitem(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    arg_check!(
-        vm,
-        args,
-        required = [
-            (dict, Some(vm.ctx.dict_type())),
-            (needle, Some(vm.ctx.str_type()))
-        ]
-    );
-
-    // What we are looking for:
-    let needle = objstr::get_value(&needle);
+        }
+    }
 
-    let elements = get_elements(dict);
-    if elements.contains_key(&needle) {
-        Ok(elements[&needle].1.clone())
-    } else {
-        Err(vm.new_value_error(format!("Key not found: {}", needle)))
+    fn setitem(self, needle: PyObjectRef, value: PyObjectRef, _vm: &mut VirtualMachine) {
+        let mut elements = self.entries.borrow_mut();
+        set_item_in_content(&mut elements, &needle, &value)
     }
-}
 
-fn dict_get(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    arg_check!(
-        vm,
-        args,
-        required = [
-            (dict, Some(vm.ctx.dict_type())),
-            (key, Some(vm.ctx.str_type()))
-        ],
-        optional = [(default, None)]
-    );
+    fn getitem(self, key: PyStringRef, vm: &mut VirtualMachine) -> PyResult {
+        let ref key = key.value;
 
-    // What we are looking for:
-    let key = objstr::get_value(&key);
+        // What we are looking for:
+        let elements = self.entries.borrow();
+        if elements.contains_key(key) {
+            Ok(elements[key].1.clone())
+        } else {
+            Err(vm.new_value_error(format!("Key not found: {}", key)))
+        }
+    }
 
-    let elements = get_elements(dict);
-    if elements.contains_key(&key) {
-        Ok(elements[&key].1.clone())
-    } else if let Some(value) = default {
-        Ok(value.clone())
-    } else {
-        Ok(vm.get_none())
+    fn get(
+        self,
+        key: PyStringRef,
+        default: OptionalArg<PyObjectRef>,
+        vm: &mut VirtualMachine,
+    ) -> PyObjectRef {
+        // What we are looking for:
+        let ref key = key.value;
+
+        let elements = self.entries.borrow();
+        if elements.contains_key(key) {
+            elements[key].1.clone()
+        } else {
+            match default {
+                OptionalArg::Present(ref value) => value.clone(),
```

Comment:
you probably don't need to `clone` here:
`OptionalArg::Present(value) => value,`